tileview/src/main/java/com/qozix/tileview/tiles/TileRenderPoolExecutor.java

```diff
@@ -0,0 +1,115 @@
+package com.qozix.tileview.tiles;
+
+import java.lang.ref.WeakReference;
+import java.util.LinkedList;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class TileRenderPoolExecutor {
+  // Sets the amount of time an idle thread will wait for a task before terminating
+  private static final int KEEP_ALIVE_TIME = 5;
+  private static final TimeUnit KEEP_ALIVE_TIME_UNIT;
+
+  private static final int CORE_POOL_SIZE = Runtime.getRuntime().availableProcessors();
+  private static final int MAXIMUM_POOL_SIZE = Runtime.getRuntime().availableProcessors();
+
+  private final CustomThreadPoolExecutor mExecutor;
+  private final BlockingQueue<Runnable> mQueue;
+  private final BlockingQueue<Future> mFutureList;
+  private WeakReference<TileCanvasViewGroup> mViewGroup;
+
+  private static TileRenderPoolExecutor sInstance = null;
+
+  static {
+    KEEP_ALIVE_TIME_UNIT = TimeUnit.SECONDS;
+    sInstance = new TileRenderPoolExecutor();
+  }
+
+  public TileRenderPoolExecutor() {
+    mQueue = new LinkedBlockingDeque<>();
+    mExecutor = new CustomThreadPoolExecutor( CORE_POOL_SIZE, MAXIMUM_POOL_SIZE, KEEP_ALIVE_TIME, KEEP_ALIVE_TIME_UNIT, mQueue );
+    mFutureList = new LinkedBlockingDeque<>();
+  }
+
+  public static TileRenderPoolExecutor getsInstance() {
+    return sInstance;
+  }
+
+  public void cancel() {
+    synchronized ( this ) {
+      for ( Future f : mFutureList ) {
+        if( !f.isDone() ){
+          f.cancel( true );
+        }
+      }
+      mQueue.clear();
+      mFutureList.clear();
+    }
+  }
+
+  public void queue( TileCanvasViewGroup viewGroup, LinkedList<Tile> tiles ) {
+    if( tiles != null && tiles.size() > 0) {
+      mViewGroup = new WeakReference<>( viewGroup );
+      viewGroup.onRenderTaskPreExecute();
+      int size = tiles.size();
+      for( int i = 0 ; i < size ; i++ ) {
+        mFutureList.add( mExecutor.submit( new TileRenderRunnable( viewGroup, tiles.get( i ) ) ) );
+      }
+    }
+  }
+
+  static class TileRenderRunnable implements Runnable {
+    private final WeakReference<TileCanvasViewGroup> mTileCanvasViewGroup;
+    private final WeakReference<Tile> mTile;
+
+    public TileRenderRunnable( TileCanvasViewGroup viewGroup, Tile tile) {
+      this.mTileCanvasViewGroup = new WeakReference<>( viewGroup );
+      this.mTile = new WeakReference<>( tile );
+    }
+
+    @Override
+    public void run() {
+      android.os.Process.setThreadPriority( android.os.Process.THREAD_PRIORITY_BACKGROUND );
+      final Thread thread = Thread.currentThread();
+
+      TileCanvasViewGroup tileCanvasViewGroup = mTileCanvasViewGroup.get();
+      Tile tile = mTile.get();
```

Comment:
nit: this should probably be moved to line 86, since if we fast-fail out on line 84, there's no reason for the fetch.  total nitpick, feel free to ignore me :)